wire-compiler/src/main/java/com/squareup/wire/java/JavaCompiler.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.java;
+
+import com.google.common.collect.ImmutableMap;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.ParameterizedTypeName;
+import com.squareup.javapoet.TypeName;
+import com.squareup.wire.ExtendableMessage;
+import com.squareup.wire.Message;
+import com.squareup.wire.internal.Util;
+import com.squareup.wire.model.ProtoTypeName;
+import com.squareup.wire.model.WireEnum;
+import com.squareup.wire.model.WireEnumConstant;
+import com.squareup.wire.model.WireOption;
+import com.squareup.wire.model.WireProtoFile;
+import com.squareup.wire.model.WireType;
+import java.util.List;
+import java.util.Map;
+import okio.ByteString;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Converts proto definitions into Java source code.
+ *
+ * <p>This can map type names from protocol buffers (like {@code uint32}, {@code string}, or {@code
+ * squareup.protos.person.Person} to the corresponding Java names (like {@code int}, {@code
+ * java.lang.String}, or {@code com.squareup.protos.person.Person}).
+ */
+public final class JavaCompiler {
```

Comment:
It also currently has the semantics of Proto2. I'm not sure if for supporting Proto3 we'll split the writers or just introduce conditional logic throughout, but something to think about.